guides/plugins/plugins/checkout/cart/tax-provider.md

```diff
@@ -0,0 +1,213 @@
+            <tag name="shopware.tax.provider" />
+        </service>
+    </services>
+```
+
+{% endcode %}
+
+## Migrate your tax provider to the database
+
+### Via migration
+
+Last but not least, you will have to create a migration to add your tax provider to the database. This is necessary, so that Shopware knows of your tax provider in the first place during checkout.
+
+{% code title="<plugin root>/src/Migration/MigrationTaxProvider.php" %}
+
+```php
+<?php declare(strict_types=1);
+
+namespace SwagTaxProviders\Migration;
+
+use Doctrine\DBAL\Connection;
+use Shopware\Core\Defaults;
+use Shopware\Core\Framework\Migration\MigrationStep;
+use Shopware\Core\Framework\Uuid\Uuid;
+use Shopware\Core\System\TaxProvider\Aggregate\TaxProviderTranslation\TaxProviderTranslationDefinition;
+use Shopware\Core\System\TaxProvider\TaxProviderDefinition;
+
+class MigrationTaxProvider extends MigrationStep
+{
+    public function getCreationTimestamp(): int
+    {
+        return 1662366151;
+    }
+
+    public function update(Connection $connection): void
+    {
+        $ruleId = $connection->fetchOne(
+            'SELECT `id` FROM `rule` WHERE `name` = :name',
+            ['name' => 'Always valid (Default)']
+        );
+
+        $id = Uuid::randomBytes();
+
+        $connection->insert(TaxProviderDefinition::ENTITY_NAME, [
+            'id' => $id,
+            'identifier' => Swag\BasicExample\Checkout\Cart\Tax\TaxProvider::class, // your tax provider class
+            'active' => true,
+            'priority' => 1,
+            'availability_rule_id' => $ruleId,
+            'created_at' => (new \DateTime())->format(Defaults::STORAGE_DATE_TIME_FORMAT),
+        ]);
+    }
+
+    public function updateDestructive(Connection $connection): void
+    {
+    }
+}
+```
+
+{% endcode %}
+
+### Via repository
+
+If you don't want to use migrations, you can also create your tax provider via repository. A good place for doing so would be the `install` lifecycle method of your plugin.
+
+If you do not know of plugin lifecycle methods yet, please refer to our [plugin lifecycle guide](../../plugin-fundamentals/plugin-lifecycle.md).
+
+{% code title="<plugin root>/src/BasicExample.php" %}
+
+```php
+<?php declare(strict_types=1);
+
+namespace SwagBasicExample;
+
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\EqualsFilter;
+use Shopware\Core\Framework\Plugin;
+use Shopware\Core\Framework\Plugin\Context\InstallContext;
+use Shopware\Core\Framework\Uuid\Uuid;
+
+class SwagBasicExample extends Plugin
+{
+    public function install(InstallContext $installContext): void
+    {
+        $ruleRepo = $this->container->get('rule.repository');
+
+        $criteria = new Criteria();
+        $criteria->addFilter(new EqualsFilter('name', 'Always valid (Default)'));
+
+        $rule = $ruleRepo->search($criteria, $installContext->getContext())->first();
```

Comment:
I would not trust this to work. This will be the example for everyone copying the code and I know that we have multiple projects without this rule name. Maybe rather suggest to add a new rule with the same structure. I'd rather have a rule, that I can just edit and only influence my tax provider, than added a new rule and have to assign it later to the tax provider. Now I question myself how this is solved for apps.